syntax = "proto3"; 
import "google/protobuf/any.proto";

package gtpo.pb;

message Node {
    int32   node_id  = 1;
    string  label    = 2;
    double  x        = 3;
    double  y        = 4;
    double  z        = 5;
    double  width    = 6;
    double  height   = 7;
    int32	group_id = 8;

    google.protobuf.Any user1   = 42;  // Reserved for user extensions
}

message Edge {
    int32   edge_id     = 1;
    int32   src_node_id = 2;
    int32   dst_node_id = 3;
    double  weight      = 4;
    double  z           = 5;
    int32   dst_edge_id = 6;

    google.protobuf.Any user1   = 42;  // Reserved for user extensions
}

message Group {
    int32   group_id            = 1;
    string  label               = 2;
    double  x                   = 3;
    double  y                   = 4;
    double  z                   = 5;
    double  width               = 6;
    double  height              = 7;

    repeated int32 node_ids     = 12;

    google.protobuf.Any user1   = 42;  // Reserved for user extensions
}

message Void {	// Empty message type serialized by default to GTpoGraph.user1 and  GTpoGraph.user2

}

message Graph {
    int32 node_count                    = 1;
    int32 edge_count                    = 2;
    int32 group_count                   = 3;

    repeated google.protobuf.Any nodes  = 12;
    repeated google.protobuf.Any edges  = 13;
    repeated google.protobuf.Any groups = 14;
}
